storage: add cache target estimates and space management control loop skeleton

dotnwat opened the pull request:
Adds a cloud cache interface for estimating the target minimum for storage size, as well as a control loop skeleton which will be used to manage disk space to be added in subsequent PR.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

dotnwat commented:
Force-push:

* Fixed circular dependency in the shared library build.

andrwng commented on src/v/config/configuration.cc:
```diff
@@ -1840,6 +1840,13 @@ configuration::configuration()
       "the data directory. Redpanda will refuse to start if it is not found.",
       {.needs_restart = needs_restart::no, .visibility = visibility::user},
       false)
+  , enable_storage_space_manager(
+      *this,
+      "enable_storage_space_manager",
+      "Enable the storage space manager that coordinates and control space "
+      "usage between log data and the cloud storage cache.",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      true)
```
I don't think the behavior you're implementing warrant doing anything differently in the face of skew, but I'm curious if you've given thought to whether this should be configurable/controllable at the node level. A few scenarios come to mind when space usage may vary across nodes
- when there is some hypothetical retention bug that consumes the space on a a single node but not others
- when there is heterogeneous hardware
- when there is severe partition-size skew

Just thinking whether during an incident it'll be desirable to have a way to flip this off for one but not all nodes.

dotnwat replied:
this makes a lot of sense. i'd even assume that we'd want to be able to turn it off on a per node basis without needing a restart. at least i think that we dont' allow that later feature so we'd need an admin endpoint.

dotnwat replied:
I've added a ticket to track this here. It's a good idea: https://github.com/redpanda-data/redpanda/issues/11192

andrwng reviewed (approved):
LGTM, nothing blocking

dotnwat commented on src/v/resource_mgmt/storage.cc:
```diff
@@ -11,21 +11,124 @@
+    co_await _gate.close();
+}
+
+ss::future<> disk_space_manager::controller() {
```
yep. fixed

dotnwat commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -1101,4 +1103,79 @@ materialized_segments& remote_partition::materialized() {
     return _api.materialized();
 }
 
+cache_usage remote_partition::get_cache_usage() const {
```
done!

dotnwat commented:
Force push:
* Fix naming of usage/target per Johns suggestion
* Fixed naming of control loop per Andrews suggestion
* Updated tests to handle recent v2/v3 changes from Abhijat

dotnwat commented:
Failures
* https://github.com/redpanda-data/redpanda/issues/11166
* https://github.com/redpanda-data/redpanda/issues/11182
* https://github.com/redpanda-data/redpanda/issues/9755
* https://github.com/redpanda-data/redpanda/issues/11169
* https://github.com/redpanda-data/redpanda/issues/8217

dotnwat commented:
Ping @andrwng

andrwng reviewed (approved):
LGTM!
